File: app/src/main/java/com/elegion/android/ui/base/fragment/BaseRecyclerFragment.java
package com.elegion.android.ui.base.fragment;

import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.elegion.android.R;
import com.elegion.android.ui.widget.EmptyView;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Mike
 */
public abstract class BaseRecyclerFragment extends BaseRefresherFragment {
    @BindView(R.id.recycler)
    protected RecyclerView mRecyclerView;
    @BindView(R.id.empty_view)
    protected EmptyView mEmptyView;

    protected RecyclerView.LayoutManager mLayoutManager;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(getLayout(), container, false);
        mUnbinder = ButterKnife.bind(this, view);
        initRecycler();
        return view;
    }

    @Override
    protected boolean shouldShowNoInternetStubView() {
        return mRecyclerView.getAdapter().getItemCount() == 0;
    }

    protected int getLayout() {
        return R.layout.fr_recycler;
    }

    @CallSuper
    protected void initRecycler() {
        mLayoutManager = getLayoutManager();
        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.setAdapter(getAdapter());
    }

    @NonNull
    protected RecyclerView.LayoutManager getLayoutManager() {
        return new LinearLayoutManager(getActivity());
    }

    protected abstract RecyclerView.Adapter getAdapter();

    protected void bindEmptyView(@DrawableRes int iconRes, @StringRes int msgTextRes,
                                 @StringRes int buttonTextRes, View.OnClickListener listener) {
        mEmptyView.bindEmptyView(iconRes, msgTextRes, buttonTextRes, listener);
    }

    protected void bindEmptyView(@DrawableRes int iconRes, @StringRes int msgTextRes) {
        mEmptyView.bindEmptyView(iconRes, msgTextRes);
    }

    public void bindEmptyView(@StringRes int text) {
        mEmptyView.bindEmptyView(text);
    }

    @Override
    protected View[] getViews() {
        return new View[]{mRecyclerView};
    }
}

File: app/src/main/java/com/elegion/android/template/ui/base/fragment/BaseRecyclerFragment.kt
package com.elegion.android.template.ui.base.fragment

import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.annotation.DrawableRes
import android.support.annotation.LayoutRes
import android.support.annotation.StringRes
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.elegion.android.template.R
import kotlinx.android.synthetic.main.fr_recycler.*

abstract class BaseRecyclerFragment : BaseRefresherFragment() {
    open val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(activity)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(getLayout(), container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initRecycler()
    }

    override fun shouldShowNoInternetStubView(): Boolean = recycler.adapter.itemCount == 0

    @LayoutRes
    override fun getLayout(): Int = R.layout.fr_recycler

    @CallSuper
    protected fun initRecycler() {
        recycler.layoutManager = layoutManager
        recycler.adapter = getAdapter()
    }

    protected abstract fun getAdapter(): RecyclerView.Adapter<*>

    protected fun bindEmptyView(@DrawableRes iconRes: Int, @StringRes msgTextRes: Int,
                                @StringRes buttonTextRes: Int, listener: View.OnClickListener) =
            emptyView.bindEmptyView(iconRes, msgTextRes, buttonTextRes, listener)

    protected fun bindEmptyView(@DrawableRes iconRes: Int, @StringRes msgTextRes: Int) =
            emptyView.bindEmptyView(iconRes, msgTextRes)

    fun bindEmptyView(@StringRes text: Int) = emptyView.bindEmptyView(text)

    override fun getViews(): Array<View> = arrayOf(recycler)
}
